presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java

```diff
@@ -444,32 +444,28 @@ private RelationPlan createTableWriterPlan(
 
             TableStatisticAggregation result = statisticsAggregationPlanner.createStatisticsAggregation(statisticsMetadata, columnToVariableMap);
 
-            StatisticAggregations.Parts aggregations = result.getAggregations().createPartialAggregations(variableAllocator, metadata.getFunctionManager());
-
-            // partial aggregation is run within the TableWriteOperator to calculate the statistics for
-            // the data consumed by the TableWriteOperator
-            // final aggregation is run within the TableFinishOperator to summarize collected statistics
-            // by the partial aggregation from all of the writer nodes
-            StatisticAggregations partialAggregation = aggregations.getPartialAggregation();
-
-            PlanNode writerNode = new TableWriterNode(
-                    idAllocator.getNextId(),
-                    source,
-                    target,
-                    variableAllocator.newVariable("partialrows", BIGINT),
-                    variableAllocator.newVariable("fragment", VARBINARY),
-                    variableAllocator.newVariable("tablecommitcontext", VARBINARY),
-                    plan.getFieldMappings(),
-                    columnNames,
-                    partitioningScheme,
-                    Optional.of(partialAggregation));
+            StatisticAggregations.Parts aggregations = result.getAggregations().splitIntoPartialAndFinal(variableAllocator, metadata.getFunctionManager());
 
             TableFinishNode commitNode = new TableFinishNode(
                     idAllocator.getNextId(),
-                    writerNode,
+                    new TableWriterNode(
+                            idAllocator.getNextId(),
+                            source,
+                            target,
+                            variableAllocator.newVariable("rows", BIGINT),
+                            variableAllocator.newVariable("fragments", VARBINARY),
+                            variableAllocator.newVariable("commitcontext", VARBINARY),
+                            plan.getFieldMappings(),
+                            columnNames,
+                            partitioningScheme,
+                            // partial aggregation is run within the TableWriteOperator to calculate the statistics for
+                            // the data consumed by the TableWriteOperator
+                            Optional.of(aggregations.getPartialAggregation())),
                     target,
                     variableAllocator.newVariable("rows", BIGINT),
-                    Optional.of(aggregations.getFinalAggregation()),
+                    // final aggregation is run within the TableFinishOperator to summarize collected statistics
+                    // by the partial aggregation from all of the writer nodes
+                    Optional.of(aggregations.getFinalOrIntermediateAggregation()),
```

Comment:
This is a little bit confusing. Can we have both `getFinalAggregation` and `getIntermediateAggregation`, and do `checkState` inside to make sure the right one is called? Or we have such use case where one want to get this aggregation regardless of whether it is final or intermediate?

Reply:
Refactored into something like https://gist.github.com/arhimondr/6563d1dcd0ad084c25ebd67c3774be89